waiter/src/waiter/token.clj

```diff
@@ -719,3 +725,119 @@
+                                                  (async/tap tokens-event-mult watch-chan)
+                                                  (update current-state :watches-count inc))
+                                           :remove (update current-state :watches-count dec)
+                                           (throw (ex-info "Invalid tokens-watch-channels-update-chan event" {:event msg})))]
+                          (log/info "Tokens-watch-maintainer watch count changed!" {:previous-count watches-count
+                                                                                    :new-count (get next-state :watches-count)})
+                          [next-state watch-refresh-timeout-chan]))
+
+                      watch-refresh-timeout-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "tokens-watch-maintainer" "refresh")
+                        (let [next-token->index (get-token->token-index-fn)
+                              [only-old-indexes only-next-indexes _] (data/diff token->token-index next-token->index)]
+                          ; if token in old-indexes and not in only-next-indexes, then those token indexes were deleted
+                          (doseq [[token {:keys [owner]}] only-old-indexes]
+                            (when-not (contains? only-next-indexes token)
+                              (async/>! tokens-event-chan (make-index-event "DELETE" {:owner owner :token token}))))
+                          ; if token in only-next-indexes, it has been updated (soft-delete included)
+                          (doseq [[token _] only-next-indexes]
+                            (async/>! tokens-event-chan (make-index-event "UPDATE" (get next-token->index token))))
+                          (if (and (nil? only-old-indexes) (nil? only-next-indexes))
+                            (log/info "Tokens-watch-maintainer found no differences in kv-store and current-state")
+                            (log/info "Tokens-watch-maintainer found some differences in kv-store and current-state"
+                                      {:only-in-current only-old-indexes :only-in-next only-next-indexes}))
+                          [(assoc current-state :token->token-index next-token->index)
+                           (async/timeout watch-refresh-timeout-ms)]))
+
+                      query-chan
+                      (let [response-chan msg]
+                        (async/put! response-chan current-state)
+                        [current-state watch-refresh-timeout-chan]))]
+                (if (and next-state next-watch-refresh-timeout-chan)
+                  (recur next-state next-watch-refresh-timeout-chan)
+                  (log/info "Stopping tokens-watch-maintainer as next loop values are nil"))))
+            (catch Exception e
+              (log/error e "Fatal error in tokens-watch-maintainer"))))]
```

Comment:
This is not fatal since we are not exiting the router. We need to invoke the JVM exit method.